Middle manager rolling upgrade issue with TLS support

Issue first reported here by @erikdubbelboer:
https://github.com/druid-io/druid/issues/4536#issuecomment-316820130

If I have a 0.10.0 cluster running, and I upgrade only the middleManager to a version of Druid that includes #4270, batch ingestion tasks fail with the following exception:

```
2017-07-21T01:29:16,989 WARN [main] io.druid.indexing.common.actions.RemoteTaskActionClient - Exception submitting action for task[index_wikiticker_2017-07-21T01:28:40.434Z]
java.io.IOException: Failed to locate service uri
	at io.druid.indexing.common.actions.RemoteTaskActionClient.submit(RemoteTaskActionClient.java:94) [druid-indexing-service-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.indexing.common.task.IndexTask.isReady(IndexTask.java:159) [druid-indexing-service-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.indexing.worker.executor.ExecutorLifecycle.start(ExecutorLifecycle.java:169) [druid-indexing-service-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_51]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_51]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_51]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_51]
	at io.druid.java.util.common.lifecycle.Lifecycle$AnnotationBasedHandler.start(Lifecycle.java:364) [java-util-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.java.util.common.lifecycle.Lifecycle.start(Lifecycle.java:263) [java-util-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.guice.LifecycleModule$2.start(LifecycleModule.java:156) [druid-api-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.cli.GuiceRunnable.initLifecycle(GuiceRunnable.java:101) [druid-services-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.cli.CliPeon.run(CliPeon.java:283) [druid-services-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.cli.Main.main(Main.java:108) [druid-services-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
Caused by: java.lang.NullPointerException
	at io.druid.curator.discovery.ServerDiscoverySelector$1.apply(ServerDiscoverySelector.java:60) ~[druid-server-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.curator.discovery.ServerDiscoverySelector$1.apply(ServerDiscoverySelector.java:52) ~[druid-server-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.curator.discovery.ServerDiscoverySelector.pick(ServerDiscoverySelector.java:108) ~[druid-server-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.indexing.common.actions.RemoteTaskActionClient.getServiceInstance(RemoteTaskActionClient.java:164) ~[druid-indexing-service-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.indexing.common.actions.RemoteTaskActionClient.submit(RemoteTaskActionClient.java:89) ~[druid-indexing-service-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	... 12 more
```

This is caused by `instance.getSslPort()` returning null:

```
      final int port = instance.getSslPort() >= 0 ? instance.getSslPort() : instance.getPort();
```

--- a/server/src/main/java/io/druid/curator/discovery/ServerDiscoverySelector.java
+++ b/server/src/main/java/io/druid/curator/discovery/ServerDiscoverySelector.java
@@ -54,11 +54,18 @@ public class ServerDiscoverySelector implements DiscoverySelector<Server>
     public Server apply(final ServiceInstance instance)
     {
       Preconditions.checkState(
-          instance.getPort() >= 0 || instance.getSslPort() >= 0,
+          instance.getPort() >= 0 || (instance.getSslPort() != null && instance.getSslPort() >= 0),
           "WTH?! Both port and sslPort not set"
       );
-      final int port = instance.getSslPort() >= 0 ? instance.getSslPort() : instance.getPort();
-      final String scheme = instance.getSslPort() >= 0 ? "https" : "http";
+      final int port;
+      final String scheme;
+      if (instance.getSslPort() == null) {
+        port = instance.getPort();
+        scheme = "http";
+      } else {
+        port = instance.getSslPort() >= 0 ? instance.getSslPort() : instance.getPort();
+        scheme = instance.getSslPort() >= 0 ? "https" : "http";
+      }
       return new Server()
       {
         @Override

--- a/server/src/test/java/io/druid/curator/discovery/ServerDiscoverySelectorTest.java
+++ b/server/src/test/java/io/druid/curator/discovery/ServerDiscoverySelectorTest.java
@@ -38,6 +38,7 @@ public class ServerDiscoverySelectorTest
   private ServerDiscoverySelector serverDiscoverySelector;
   private ServiceInstance instance;
   private static final int PORT = 8080;
+  private static final int SSL_PORT = 8280;
   private static final String ADDRESS = "localhost";
 
   @Before
@@ -77,6 +78,63 @@ public class ServerDiscoverySelectorTest
     Assert.assertEquals("http", uri.getScheme());
   }
 
+  @Test
+  public void testPickWithNullSslPort() throws Exception
+  {
+    EasyMock.expect(serviceProvider.getInstance()).andReturn(instance).anyTimes();
+    EasyMock.expect(instance.getAddress()).andReturn(ADDRESS).anyTimes();
+    EasyMock.expect(instance.getPort()).andReturn(PORT).anyTimes();
+    EasyMock.expect(instance.getSslPort()).andReturn(null).anyTimes();
+    EasyMock.replay(instance, serviceProvider);
+    Server server = serverDiscoverySelector.pick();
+    Assert.assertEquals(PORT, server.getPort());
+    Assert.assertEquals(ADDRESS, server.getAddress());
+    Assert.assertTrue(server.getHost().contains(Integer.toString(PORT)));
+    Assert.assertTrue(server.getHost().contains(ADDRESS));
+    Assert.assertEquals("http", server.getScheme());
+    EasyMock.verify(instance, serviceProvider);
+    final URI uri = new URI(
+        server.getScheme(),
+        null,
+        server.getAddress(),
+        server.getPort(),
+        "/druid/indexer/v1/action",
+        null,
+        null
+    );
+    Assert.assertEquals(PORT, uri.getPort());
+    Assert.assertEquals(ADDRESS, uri.getHost());
+    Assert.assertEquals("http", uri.getScheme());
+  }
+
+  @Test
+  public void testPickWithSslPort() throws Exception
+  {
+    EasyMock.expect(serviceProvider.getInstance()).andReturn(instance).anyTimes();
+    EasyMock.expect(instance.getAddress()).andReturn(ADDRESS).anyTimes();
+    EasyMock.expect(instance.getPort()).andReturn(PORT).anyTimes();
+    EasyMock.expect(instance.getSslPort()).andReturn(SSL_PORT).anyTimes();
+    EasyMock.replay(instance, serviceProvider);
+    Server server = serverDiscoverySelector.pick();
+    Assert.assertEquals(SSL_PORT, server.getPort());
+    Assert.assertEquals(ADDRESS, server.getAddress());
+    Assert.assertTrue(server.getHost().contains(Integer.toString(SSL_PORT)));
+    Assert.assertTrue(server.getHost().contains(ADDRESS));
+    Assert.assertEquals("https", server.getScheme());
+    EasyMock.verify(instance, serviceProvider);
+    final URI uri = new URI(
+        server.getScheme(),
+        null,
+        server.getAddress(),
+        server.getPort(),
+        "/druid/indexer/v1/action",
+        null,
+        null
+    );
+    Assert.assertEquals(SSL_PORT, uri.getPort());
+    Assert.assertEquals(ADDRESS, uri.getHost());
+    Assert.assertEquals("https", uri.getScheme());
+  }
 
   @Test
   public void testPickIPv6() throws Exception

